server/game/gameSystems/MoveCardSystem.ts

```diff
@@ -24,76 +32,86 @@ export class MoveCardSystem<TContext extends AbilityContext = AbilityContext> ex
-            [properties.target, destinationController, properties.destination]
+            'move {0} to ' + (properties.bottom ? 'the bottom of ' : '') + 'their {1}',
+            [properties.target, properties.destination]
         ];
     }
 
+    protected override updateEvent(event, card: Card, context: TContext, additionalProperties): void {
+        // Check if the card is leaving play
+        if (EnumHelpers.isArena(card.location) && !EnumHelpers.isArena(event.destination)) {
+            this.addLeavesPlayPropertiesToEvent(event, card, context, additionalProperties);
+        } else {
+            super.updateEvent(event, card, context, additionalProperties);
+        }
+    }
+
     public override addPropertiesToEvent(event: any, card: Card, context: TContext, additionalProperties?: any): void {
         const properties = this.generatePropertiesFromContext(context, additionalProperties);
         super.addPropertiesToEvent(event, card, context, additionalProperties);
 
-        event.switch = properties.switch;
-        event.switchTarget = properties.switchTarget;
-        event.changePlayer = properties.changePlayer;
         event.destination = properties.destination;
-        event.bottom = properties.bottom;
         event.shuffle = properties.shuffle;
+        event.options = { bottom: !!properties.bottom };
     }
 
-    public override canAffect(card: Card, context: TContext, additionalProperties = {}): boolean {
-        const { changePlayer, destination } = this.generatePropertiesFromContext(context, additionalProperties) as IMoveCardProperties;
-        return (
-            (!changePlayer ||
-              (!card.hasRestriction(EffectName.TakeControl, context) &&
-                !card.anotherUniqueInPlay(context.player))) &&
-                (!destination || context.player.isLegalLocationForCardType(card.type, destination)) &&
-                !EnumHelpers.isArena(card.location) &&
-                super.canAffect(card, context)
-        );
+    public override canAffect(card: Card, context: TContext, additionalProperties = {}, mustChangeGameState = GameStateChangeRequired.None): boolean {
+        const { destination } = this.generatePropertiesFromContext(context, additionalProperties) as IMoveCardProperties;
+
+        // Ensure that we have a valid destination and that the card can be moved there
+        if (!destination || !context.player.isLegalLocationForCardType(card.type, destination)) {
+            return false;
+        }
+
+        // Ensure that if the card is returning to the hand, it must be in the discard pile or in play or be a resource
+        if (destination === Location.Hand && !EnumHelpers.cardLocationMatches(card.location, Location.Discard) && !EnumHelpers.cardLocationMatches(card.location, Location.Resource) && !EnumHelpers.isArena(card.location)) {
+            Contract.fail(`Cannot use MoveCardSystem to return a card to hand from ${card.location}`);
```

Comment:
Instead of having an if / else that calls Contract.fail you can just use one of the assertion forms that tests a condition:

```typescript
Contract.assertFalse(destination === Location.Hand && <rest of statement>, `Cannot use MoveCardSystem to return a card to hand from ${card.location});
```